Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs

```diff
@@ -0,0 +1,162 @@
+        /// or from multiple repository objects (simulating multiple servers all starting up at the same time).
+        /// While it would be great for this to work for much larger numbers than are tested here, the GCS
+        /// limit of "one update per second, for each object" limits how far this can reliably be tested. But
+        /// it would be *very* unusual for this to actually cause a problem.
+        /// </summary>
+        [Theory]
+        [InlineData(5, 1)]
+        [InlineData(1, 5)]
+        [InlineData(2, 3)]
+        public async Task ConcurrencyStress(int repositories, int elementsPerRepository)
+        {
+            var client = StorageClient.Create();
+            string objectName = $"concurrency-{repositories}-{elementsPerRepository}.xml";
+            var semaphore = new SemaphoreSlim(0, repositories * elementsPerRepository);
+            // Start all the tasks, but they won't do anything until we release the semaphore
+            var tasks =
+                (from repoIndex in Enumerable.Range(0, repositories)
+                 let repo = CreateRepo(objectName)
+                 from elementIndex in Enumerable.Range(0, elementsPerRepository)
+                 let element = CreateElement($"{repoIndex}/{elementIndex}")
+                 select StoreAsync(repo, element, semaphore)).ToList();
+
+            semaphore.Release(repositories * elementsPerRepository);
+            await Task.WhenAll(tasks);
+            var expectedIds = from repoIndex in Enumerable.Range(0, repositories)
+                              from elementIndex in Enumerable.Range(0, elementsPerRepository)
+                              select $"{repoIndex}/{elementIndex}";
+
+            ValidateStoredDocument(objectName, expectedIds);
+        }
+
+        /// <summary>
+        /// Checks that the document stored in the given object has elements with the given IDs (and only those),
+        /// in any order.
+        /// </summary>
+        private void ValidateStoredDocument(string objectName, IEnumerable<string> expectedIds)
+        {
+            var client = StorageClient.Create();
+            XDocument document;
+            using (var stream = new MemoryStream())
+            {
+                client.DownloadObject(_fixture.Bucket, objectName, stream);
+                stream.Position = 0;
+                document = XDocument.Load(stream);
+            }
+            var allIds = document.Root.Elements().Select(x => x.Attribute("id").Value).OrderBy(id => id).ToList();
+            expectedIds = expectedIds.OrderBy(id => id).ToList();
+            Assert.Equal(expectedIds, allIds);
+        }
+
+        private XElement CreateElement(string id) => new XElement("element", new XAttribute("id", id));
+
+        /// <summary>
+        /// Starts a task which will wait for the given semaphore to be released, then
+        /// adds the given item to the given repository. This uses a thread rather than a task to ensure
```

Comment:
I don't understand how this is ensuring that a single thread isn't being used?

Reply:
Comment was written when it was going to return Thread. I fail.
Have added a ConfigureAwait(false) so things really will be running on the thread pool. We *could* use Task.Factory.StartNew with a "long-running", but that's probably overkill.